### ISSUE TITLE
VIP and firewall policy destroy\update order dependency

### ISSUE THREAD

**Issue by murakest** [fortinetdev/terraform-provider-fortios#123]
Hello,
Our enviroment is
FortiOS version: v6.4.2 build1723 (GA)
Terraform version: 0.14.2
Fortios provider version: 1.6.16

There is a problem during VIP name update if it is being used by firewall policy.
What happens is that terraform wants to re-create VIP and only update in place firewall policy. Terraform fails as you can't destroy VIP being used.

create_before_destroy = true also does not help as terraform can't create another VIP as the same mapped IP is already in use.

In this situation fw policy should be destroyed first, then vip destroyed and everything recreated vice versa.

```
terraform {
  required_providers {
    fortios = {
      source = "fortinetdev/fortios"
    }
  }
}

locals {
  services = split(",", upper(var.fw_policy_services))
}

resource "fortios_firewall_vip" "vapp_vip" {
  mappedip {
    range = var.vip_mappedip
  }
  name        = var.vip_name
  comment     = var.vip_comment
  extintf     = var.vip_extintf
  type        = var.vip_type
  extip       = var.vip_extip
  portforward = var.vip_portforward ? "enable" : "disable"
  protocol    = var.vip_portforward ? var.vip_protocol : null
  extport     = var.vip_portforward ? var.vip_portforward_extport : null
  mappedport  = var.vip_portforward ? var.vip_portforward_mappedport : null

  lifecycle {
    create_before_destroy = true
  }
}

resource "fortios_firewall_policy" "vapp_firewall_policy" {
  name     = var.fw_policy_name
  action   = var.fw_policy_action
  schedule = var.fw_policy_schedule
  policyid = 1 # needs different per policy

  dstaddr {
    name = fortios_firewall_vip.vapp_vip.name
  }

  dstintf {
    name = var.fw_policy_outgoing_interface
  }

  srcaddr {
    name = var.fw_policy_source
  }

  srcintf {
    name = var.fw_policy_incoming_interface
  }

  dynamic "service" {
    for_each = [for service in local.services : {
      name = service
    }]

    content {
      name = service.value.name
    }
  }
}
```
Terraform plan output
```
  # module.staticDnsNatFwPolicy[0].fortios_firewall_policy.vapp_firewall_policy will be updated in-place
  ~ resource "fortios_firewall_policy" "vapp_firewall_policy" {
        id                          = "1"
      ~ name                        = "aaaa-01" -> "aaaa-02"
        # (72 unchanged attributes hidden)

      ~ dstaddr {
          ~ name = "aaaa-01" -> "aaaa-02"
        }
        # (4 unchanged blocks hidden)
    }

  # module.staticDnsNatFwPolicy[0].fortios_firewall_vip.vapp_vip must be replaced
+/- resource "fortios_firewall_vip" "vapp_vip" {
      ~ arp_reply                        = "enable" -> (known after apply)
      ~ color                            = 0 -> (known after apply)
      ~ dns_mapping_ttl                  = 0 -> (known after apply)
      ~ extport                          = "0-65535" -> (known after apply)
      ~ fosid                            = 0 -> (known after apply)
      ~ gratuitous_arp_interval          = 0 -> (known after apply)
      ~ http_cookie_age                  = 60 -> (known after apply)
      + http_cookie_domain               = (known after apply)
      ~ http_cookie_domain_from_host     = "disable" -> (known after apply)
      ~ http_cookie_generation           = 0 -> (known after apply)
      + http_cookie_path                 = (known after apply)
      ~ http_cookie_share                = "same-ip" -> (known after apply)
      ~ http_ip_header                   = "disable" -> (known after apply)
      + http_ip_header_name              = (known after apply)
      ~ http_multiplex                   = "disable" -> (known after apply)
      ~ https_cookie_secure              = "disable" -> (known after apply)
      ~ id                               = "aaaa-01" -> (known after apply)
      ~ ldb_method                       = "static" -> (known after apply)
      + mapped_addr                      = (known after apply)
      ~ mappedport                       = "0-65535" -> (known after apply)
      ~ max_embryonic_connections        = 1000 -> (known after apply)
      ~ name                             = "aaaa-01" -> "aaaa-02" # forces replacement
      ~ nat_source_vip                   = "disable" -> (known after apply)
      ~ outlook_web_access               = "disable" -> (known after apply)
      ~ persistence                      = "none" -> (known after apply)
      ~ portmapping_type                 = "1-to-1" -> (known after apply)
      ~ protocol                         = "tcp" -> (known after apply)
      + server_type                      = (known after apply)
      ~ ssl_algorithm                    = "high" -> (known after apply)
      + ssl_certificate                  = (known after apply)
      ~ ssl_client_fallback              = "enable" -> (known after apply)
      ~ ssl_client_renegotiation         = "secure" -> (known after apply)
      ~ ssl_client_session_state_max     = 1000 -> (known after apply)
      ~ ssl_client_session_state_timeout = 30 -> (known after apply)
      ~ ssl_client_session_state_type    = "both" -> (known after apply)
      ~ ssl_dh_bits                      = "2048" -> (known after apply)
      ~ ssl_hpkp                         = "disable" -> (known after apply)
      ~ ssl_hpkp_age                     = 5184000 -> (known after apply)
      + ssl_hpkp_backup                  = (known after apply)
      ~ ssl_hpkp_include_subdomains      = "disable" -> (known after apply)
      + ssl_hpkp_primary                 = (known after apply)
      ~ ssl_hsts                         = "disable" -> (known after apply)
      ~ ssl_hsts_age                     = 5184000 -> (known after apply)
      ~ ssl_hsts_include_subdomains      = "disable" -> (known after apply)
      ~ ssl_http_location_conversion     = "disable" -> (known after apply)
      ~ ssl_http_match_host              = "enable" -> (known after apply)
      ~ ssl_max_version                  = "tls-1.3" -> (known after apply)
      ~ ssl_min_version                  = "tls-1.1" -> (known after apply)
      ~ ssl_mode                         = "half" -> (known after apply)
      ~ ssl_pfs                          = "require" -> (known after apply)
      ~ ssl_send_empty_frags             = "enable" -> (known after apply)
      ~ ssl_server_algorithm             = "client" -> (known after apply)
      ~ ssl_server_max_version           = "client" -> (known after apply)
      ~ ssl_server_min_version           = "client" -> (known after apply)
      ~ ssl_server_session_state_max     = 100 -> (known after apply)
      ~ ssl_server_session_state_timeout = 60 -> (known after apply)
      ~ ssl_server_session_state_type    = "both" -> (known after apply)
      ~ uuid                             = "7df2764a-3e2b-51eb-6eff-4d3e37f68acb" -> (known after apply)
      ~ weblogic_server                  = "disable" -> (known after apply)
      ~ websphere_server                 = "disable" -> (known after apply)
        # (4 unchanged attributes hidden)
        # (1 unchanged block hidden)
    }
```

**Comment by frankshen01** [comment]
Hi @murakest , Thanks for the details.  Feature fixed/supported and released, please see the latest version(v1.6.17): https://registry.terraform.io/providers/fortinetdev/fortios/latest  for details.  Validation:

1 terraform config

```
# cat main.tf

variable testname {
  type        = string
  default     = "111"
  description = "description"
}

resource "fortios_firewall_vip" "vapp_vip" {
  mappedip {
    range = "3.0.0.0-3.0.0.1"
  }
  name        = var.testname
  extintf     = "any"
  type        = "static-nat"
  extip       = "1.0.0.1-1.0.0.2"
  portforward = "disable"
  protocol    = "tcp"
  extport     = "0-65535"
  mappedport  = "0-65535"
}

resource "fortios_firewall_policy" "trname" {
  action             = "deny"
  name               = join("-", [var.testname, "policy"])
  schedule           = "always"
  wanopt             = "disable"
  wanopt_detection   = "active"
  wanopt_passive_opt = "default"
  wccp               = "disable"
  webcache           = "disable"
  webcache_https     = "disable"
  wsso               = "enable"

  dstaddr {
    name = fortios_firewall_vip.vapp_vip.name
  }

  dstintf {
    name = "port4"
  }

  service {
    name = "HTTP"
  }

  srcaddr {
    name = "all"
  }

  srcintf {
    name = "port3"
  }
}

```

2 terraform apply and show the result on firewall

```
mytestFortiGate (vip) # show
config firewall vip
    edit "111"
        set uuid 6230e52a-38bf-51eb-32ec-9a082bbfd9a7
        set extip 1.0.0.1-1.0.0.2
        set extintf "any"
        set mappedip "3.0.0.0-3.0.0.1"
    next
end

mytestFortiGate (vip) # end

mytestFortiGate # config firewall policy

mytestFortiGate (policy) # show
config firewall policy
    edit 1
        set name "111-policy"
        set uuid 6239f8c2-38bf-51eb-7cbb-b765aced6d98
        set srcintf "port3"
        set dstintf "port4"
        set srcaddr "all"
        set dstaddr "111"
        set schedule "always"
        set service "HTTP"
        set logtraffic disable
    next
end

```

3 rename variable's default value

```
variable testname {
  type        = string
  default     = "222"
  description = "description"
}

```

4 terraform apply

```
# terraform apply
  ----
fortios_firewall_vip.vapp_vip: Refreshing state... [id=111]
fortios_firewall_policy.trname: Refreshing state... [id=1]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # fortios_firewall_policy.trname will be updated in-place
  ~ resource "fortios_firewall_policy" "trname" {
        action                      = "deny"
        auth_path                   = "disable"
        block_notification          = "disable"
        captive_portal_exempt       = "disable"
        capture_packet              = "disable"
        delay_tcp_npu_session       = "disable"
        diffserv_forward            = "disable"
        diffserv_reverse            = "disable"
        diffservcode_forward        = "000000"
        diffservcode_rev            = "000000"
        disclaimer                  = "disable"
        dsri                        = "disable"
        dstaddr_negate              = "disable"
        firewall_session_dirty      = "check-all"
        fixedport                   = "disable"
        fsso                        = "enable"
        id                          = "1"
        inbound                     = "disable"
        internet_service            = "disable"
        internet_service_negate     = "disable"
        internet_service_src        = "disable"
        internet_service_src_negate = "disable"
        ippool                      = "disable"
        learning_mode               = "disable"
        logtraffic                  = "disable"
        logtraffic_start            = "disable"
        match_vip                   = "disable"
      ~ name                        = "111-policy" -> "222-policy"
        nat                         = "disable"
        natinbound                  = "disable"
        natip                       = "0.0.0.0 0.0.0.0"
        natoutbound                 = "disable"
        ntlm                        = "disable"
        ntlm_guest                  = "disable"
        outbound                    = "enable"
        permit_any_host             = "disable"
        permit_stun_host            = "disable"
        policyid                    = 1
        profile_protocol_options    = "default"
        profile_type                = "single"
        radius_mac_auth_bypass      = "disable"
        reputation_direction        = "destination"
        reputation_minimum          = 0
        rsso                        = "disable"
        rtp_nat                     = "disable"
        scan_botnet_connections     = "disable"
        schedule                    = "always"
        schedule_timeout            = "disable"
        send_deny_packet            = "disable"
        service_negate              = "disable"
        session_ttl                 = 0
        srcaddr_negate              = "disable"
        ssl_mirror                  = "disable"
        status                      = "enable"
        tcp_mss_receiver            = 0
        tcp_mss_sender              = 0
        tcp_session_without_syn     = "disable"
        timeout_send_rst            = "disable"
        tos                         = "0x00"
        tos_mask                    = "0x00"
        tos_negate                  = "disable"
        utm_status                  = "disable"
        uuid                        = "6239f8c2-38bf-51eb-7cbb-b765aced6d98"
        vlan_cos_fwd                = 255
        vlan_cos_rev                = 255
        wanopt                      = "disable"
        wanopt_detection            = "active"
        wanopt_passive_opt          = "default"
        wccp                        = "disable"
        webcache                    = "disable"
        webcache_https              = "disable"
        wsso                        = "enable"

      ~ dstaddr {
          ~ name = "111" -> "222"
        }

        dstintf {
            name = "port4"
        }

        service {
            name = "HTTP"
        }

        srcaddr {
            name = "all"
        }

        srcintf {
            name = "port3"
        }
    }

  # fortios_firewall_vip.vapp_vip will be updated in-place
  ~ resource "fortios_firewall_vip" "vapp_vip" {
        arp_reply                        = "enable"
        color                            = 0
        dns_mapping_ttl                  = 0
        extintf                          = "any"
        extip                            = "1.0.0.1-1.0.0.2"
        extport                          = "0-65535"
        fosid                            = 0
        gratuitous_arp_interval          = 0
        http_cookie_age                  = 60
        http_cookie_domain_from_host     = "disable"
        http_cookie_generation           = 0
        http_cookie_share                = "same-ip"
        http_ip_header                   = "disable"
        http_multiplex                   = "disable"
        https_cookie_secure              = "disable"
        id                               = "111"
        ldb_method                       = "static"
        mappedport                       = "0-65535"
        max_embryonic_connections        = 1000
      ~ name                             = "111" -> "222"
        nat_source_vip                   = "disable"
        outlook_web_access               = "disable"
        persistence                      = "none"
        portforward                      = "disable"
        portmapping_type                 = "1-to-1"
        protocol                         = "tcp"
        ssl_algorithm                    = "high"
        ssl_client_fallback              = "enable"
        ssl_client_renegotiation         = "secure"
        ssl_client_session_state_max     = 1000
        ssl_client_session_state_timeout = 30
        ssl_client_session_state_type    = "both"
        ssl_dh_bits                      = "2048"
        ssl_hpkp                         = "disable"
        ssl_hpkp_age                     = 5184000
        ssl_hpkp_include_subdomains      = "disable"
        ssl_hsts                         = "disable"
        ssl_hsts_age                     = 5184000
        ssl_hsts_include_subdomains      = "disable"
        ssl_http_location_conversion     = "disable"
        ssl_http_match_host              = "enable"
        ssl_max_version                  = "tls-1.2"
        ssl_min_version                  = "tls-1.1"
        ssl_mode                         = "half"
        ssl_pfs                          = "require"
        ssl_send_empty_frags             = "enable"
        ssl_server_algorithm             = "client"
        ssl_server_max_version           = "client"
        ssl_server_min_version           = "client"
        ssl_server_session_state_max     = 100
        ssl_server_session_state_timeout = 60
        ssl_server_session_state_type    = "both"
        type                             = "static-nat"
        uuid                             = "6230e52a-38bf-51eb-32ec-9a082bbfd9a7"
        weblogic_server                  = "disable"
        websphere_server                 = "disable"

        mappedip {
            range = "3.0.0.0-3.0.0.1"
        }
    }

Plan: 0 to add, 2 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_firewall_vip.vapp_vip: Modifying... [id=111]
fortios_firewall_vip.vapp_vip: Modifications complete after 0s [id=222]
fortios_firewall_policy.trname: Modifying... [id=1]
fortios_firewall_policy.trname: Modifications complete after 0s [id=1]

Apply complete! Resources: 0 added, 2 changed, 0 destroyed.
root@sv:/work/zamba/03a-fos-gen-6.2.0only/03Terraform/bin#

```
We can see in-place update completed.

5 show result on firewall:
```

mytestFortiGate # config firewall policy

mytestFortiGate (policy) # show
config firewall policy
    edit 1
        set name "222-policy"
        set uuid 6239f8c2-38bf-51eb-7cbb-b765aced6d98
        set srcintf "port3"
        set dstintf "port4"
        set srcaddr "all"
        set dstaddr "222"
        set schedule "always"
        set service "HTTP"
        set logtraffic disable
    next
end

mytestFortiGate (policy) # end

mytestFortiGate # config firewall vip

mytestFortiGate (vip) # show
config firewall vip
    edit "222"
        set uuid 6230e52a-38bf-51eb-32ec-9a082bbfd9a7
        set extip 1.0.0.1-1.0.0.2
        set extintf "any"
        set mappedip "3.0.0.0-3.0.0.1"
    next
end

```
Let me know if you need anything else. Thank you!

**Comment by murakest** [comment]
Works fine now. Thank you.
